amethyst_renderer/src/sprite/sprite_sheet_gen.rs

```diff
@@ -0,0 +1,354 @@
+        let padding_pixels = if params.padded { 1 } else { 0 };
+        let offset_w = params.sprite_w + padding_pixels;
+        let offset_h = params.sprite_h + padding_pixels;
+        'outer: for row in 0..params.row_count {
+            for col in 0..params.column_count {
+                let offset_x = offset_w * col as u32;
+                let offset_y = offset_h * row as u32;
+                let sprite = self.from_pixel_values(
+                    image_w as u32,
+                    image_h as u32,
+                    params.sprite_w,
+                    params.sprite_h,
+                    offset_x,
+                    offset_y,
+                    [0.; 2],
+                );
+
+                sprites.push(sprite);
+
+                if sprites.len() == sprite_count {
+                    break 'outer;
+                }
+            }
+        }
+
+        sprites
+    }
+
+    /// Creates a `Sprite` from pixel values.
+    ///
+    /// This function expects pixel coordinates -- starting from the top left of the image. X
+    /// increases to the right, Y increases downwards. Texture coordinates are calculated from the
+    /// pixel values.
+    ///
+    /// # Parameters
+    ///
+    /// * `image_w`: Width of the full sprite sheet.
+    /// * `image_h`: Height of the full sprite sheet.
+    /// * `sprite_w`: Width of the sprite.
+    /// * `sprite_h`: Height of the sprite.
+    /// * `pixel_left`: Pixel X coordinate of the left side of the sprite.
+    /// * `pixel_top`: Pixel Y coordinate of the top of the sprite.
+    /// * `offsets`: Number of pixels to shift the sprite to the left and down relative to the
+    ///              entity.
+    pub fn from_pixel_values(
+        self,
+        image_w: u32,
+        image_h: u32,
+        sprite_w: u32,
+        sprite_h: u32,
+        pixel_left: u32,
+        pixel_top: u32,
+        offsets: [f32; 2],
+    ) -> Sprite {
+        // Fraction of pixels to shift inward from the edge as texture coordinate adjustment.
+        // `0.` means texture coordinates lie exactly on the pixel edge.
+        let edge_shift = match self {
+            SpriteSheetGen::Edge => 0.,
+            SpriteSheetGen::HalfPixel => 0.5,
+        };
+
+        let image_w = image_w as f32;
+        let image_h = image_h as f32;
+        let offsets = [offsets[0] as f32, offsets[1] as f32];
+
+        let pixel_right = (pixel_left + sprite_w) as f32;
+        let pixel_bottom = (pixel_top + sprite_h) as f32;
+        let pixel_left = pixel_left as f32;
+        let pixel_top = pixel_top as f32;
+
+        // Texture coordinates are expressed as fractions of the position on the image.
+        // Y axis texture coordinates start at the bottom of the image, so we have to invert them.
+        //
+        // For pixel perfect result, the sprite border must be rendered exactly at
+        // screen pixel border or use nearest-neighbor sampling.
```

Comment:
This isn't true. This logic breaks down on non-unit display pixel ratios.

What could be instead done is to also shift `image_w` and `image_h` down, so the texture would be trimmed, but not stretched.

Reply:
hm, am having trouble understanding non-unit display pixel ratios [[1], [2]]. I need a less text, more picture explanation `@_@` (or a revisit of those pages when my brain can take it).

[1]: https://stackoverflow.com/questions/8785643/what-exactly-is-device-pixel-ratio
[2]: http://weibeld.net/webdesign/device-pixel-ratio.html

Reply:
Basically think about that this way: Your sprite has size (W,H), while your part of texture that you map to this sprite has size (W-2p, H-2p) where p is your padding. This obviously forces it to stretch. You don't see that stretch on nearest-neighbour if it happens to cover exactly (W,H) physical pixels on your display, because how else could it display it? Once you add more pixels (your display is denser than your texture), then that stretching becomes visible. To counteract that, you have to somehow prevent stretching. You can shrink down the sprite itself to prevent stretching and it should work, but then on denser displays you must expect that sprites will actually have visible space between them.

No matter how you put it, shrinking the UVs generates more problems than it solves. There is no valid way to do it without actually modifying texture or preparing your assets accordingly.

Usually what's done to counteract that is every sprite has added 1px border outside of it boundaries that matches exactly the pixels in the sprite. In case of repeating sprites, this border has "continue" the pattern, basically copying the 1px row/column from opposite side. In that case you just continue to use standard `0..1` UVs and it will all work correctly.

Reply:
Understand the shrinking's problems for an object's sprite sheet; the shrinking UVs is okay for the single-colour sprites right?
I can come up with another PR that adds the repeated pixels around each sprite to avoid shrinking the UVs